### ISSUE TITLE
是否可以实现不对齐音素，只根据频谱切分边界？标记全部用SP

### ISSUE THREAD

**Issue by ILG2021** [qiuqiao/SOFA#35]
中文的准确率还行，外文几乎没啥用，不知道跟训练数据量或者外文的音素复杂性有关，因为实际的音素会有连读，省略的情况，g2p出来的音素序列不一定准确，这在一定程度上影响模型的推理。另外少量数据好像无法实现比较高的准确率。我想使用模型只切分边界，手动填入音素。这样少量数据（30分钟左右）也可以比较精准。

**Comment by ILG2021** [comment]
已初步实现，效果还可以。只是训练过程中valid loss很不正常，一直升高，望作者指导下。
我先说下思路，原来是多分类问题，需要预测每个frame所属的phoneme，而我实现的是简化任务，就是二分类问题，在切分线所在的frame上是1，其他地方都是0。任务简化之后，模型可以更加胜任。

1. binarize中
```
ph_frame[int(np.round(st)) : int(np.round(ed))] = ph_id
```
修改为：
```
ph_frame[int(np.round(st))] = 1
ph_frame[int(np.round(ed))] = 1
```
2. unet head改成2+1，2是二分类，1是edge
3. 去掉frame_loss中的掩码，因为二分类问题每个类别都是有效的
```
ph_frame_GHM_loss = self.ph_frame_GHM_loss_fn(
            ph_frame_logits,
            ph_frame_gt,
            (mask.unsqueeze(-1) * ph_mask.unsqueeze(1)),
            valid,
        )
```
修改为：
```
ph_frame_GHM_loss = self.ph_frame_GHM_loss_fn(
            ph_frame_logits,
            ph_frame_gt,
            mask,
            valid,
        )
```
不知道这个mask是不是处理的有问题，因为维度跟之前不一样

训练过程如下：
![image](https://github.com/user-attachments/assets/697bee9e-cd4e-4e7d-a004-4f59583fd921)

**Comment by qiuqiao** [comment]
这个是可以的，SOME用的就是类似的方法，SOFA v1用这个方法进行辅助训练。SOFA v2有一个类似的CIF软对齐（还没写）。
valid loss的话，按你的描述，应该只需要EMD loss够小就行了，其他loss起的是辅助作用，貌似用处不是很大。

**Comment by ILG2021** [comment]
ph_frame_GHM_loss = self.ph_frame_GHM_loss_fn(
            ph_frame_logits,
            ph_frame_gt,
            mask,
            valid,
        )
mask的维度是(B,T)，是不是应该是(B,T,2)?

**Comment by ILG2021** [comment]
实测结果是有些是边界的地方漏切，可能是因为我这边的边界数据远小于非边界数据，这样我该如何设计loss函数呢？
